qutip/core/environment.py

```diff
@@ -0,0 +1,2015 @@
+
+        T : optional, float
+            Bath temperature. (The correlation function and the power spectrum
+            of this environment can only be calculated from the spectral
+            density if a temperature is provided.)
+
+        tag : optional, str, tuple or any other object
+            An identifier (name) for this environment.
+        """
+        return _BosonicEnvironment_fromSD(J, wlist, wMax, T, tag)
+
+    # --- spectral density, power spectrum, correlation function conversions
+
+    def _ps_from_sd(self, w, eps, derivative=None):
+        # derivative: value of J'(0)
+        if self.T is None:
+            raise ValueError(
+                "Bath temperature must be specified for this operation")
+
+        w = np.array(w, dtype=float)
+        if self.T == 0:
+            return 2 * np.heaviside(w, 0) * self.spectral_density(w)
+
+        # at zero frequency, we do numerical differentiation
+        # S(0) = 2 J'(0) / beta
+        zero_mask = (w == 0)
+        nonzero_mask = np.invert(zero_mask)
+
+        S = np.zeros_like(w)
+        if derivative is None:
+            S[zero_mask] = 2 * self.T * self.spectral_density(eps) / eps
+        else:
+            S[zero_mask] = 2 * self.T * derivative
+        S[nonzero_mask] = (
+            2 * np.sign(w[nonzero_mask])
+            * self.spectral_density(np.abs(w[nonzero_mask]))
+            * (n_thermal(w[nonzero_mask], self.T) + 1)
+        )
+        return S
+
+    def _sd_from_ps(self, w):
+        if self.T is None:
+            raise ValueError(
+                "Bath temperature must be specified for this operation")
+
+        w = np.array(w, dtype=float)
+        J = np.zeros_like(w)
+        positive_mask = (w > 0)
+
+        J[positive_mask] = (
+            self.power_spectrum(w[positive_mask]) / 2
+            / (n_thermal(w[positive_mask], self.T) + 1)
+        )
+        return J
+
+    def _ps_from_cf(self, w, tMax):
+        w = np.array(w, dtype=float)
+        if w.ndim == 0:
+            wMax = np.abs(w)
+        else:
+            wMax = max(np.abs(w[0]), np.abs(w[-1]))
+
+        mirrored_result = _fft(self.correlation_function, wMax, tMax=tMax)
+        return np.real(mirrored_result(-w))
+
+    def _cf_from_ps(self, t, wMax, **ps_kwargs):
+        t = np.array(t, dtype=float)
+        if t.ndim == 0:
+            tMax = np.abs(t)
+        else:
+            tMax = max(np.abs(t[0]), np.abs(t[-1]))
+
+        result = _fft(lambda w: self.power_spectrum(w, **ps_kwargs),
+                      tMax, tMax=wMax)
+        return result(t) / (2 * np.pi)
+
+    # --- fitting
+
+    def approx_by_cf_fit(
+        self, tlist: ArrayLike, target_rsme: float = 2e-5, Nr_max: int = 10,
+        Ni_max: int = 10, guess: list[float] = None, lower: list[float] = None,
+        upper: list[float] = None, full_ansatz: bool = False, tag: Any = None
+    ) -> tuple[ExponentialBosonicEnvironment, dict[str, Any]]:
```

Comment:
can we support passsing more options to curve_fit, particularly sigma and max_fev?

Reply:
Ah I left out that `sigma` because there were so many parameters already and I couldn't imagine it being useful, but if it is, sure, we can add it back in!

Reply:
Sure, I'll add `sigma` back, and `add max_fev`, which it's just hardcoded now :)

Reply:
Yeah, in some cases I found controlling sigma pretty useful